crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -5253,36 +5254,130 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
+    fn infer_set_expression(&mut self, set: &ast::ExprSet, tcx: TypeContext<'db>) -> Type<'db> {
         let ast::ExprSet {
             range: _,
             node_index: _,
             elts,
         } = set;
 
-        // TODO: Use the type context for more precise inference.
+        self.infer_collection_literal(elts, tcx, KnownClass::Set)
+            .unwrap_or_else(|| {
+                KnownClass::Set.to_specialized_instance(self.db(), [Type::unknown()])
+            })
+    }
+
+    // Infer the type of a collection literal expression.
+    fn infer_collection_literal(
+        &mut self,
+        elts: &[ast::Expr],
+        tcx: TypeContext<'db>,
+        collection_class: KnownClass,
+    ) -> Option<Type<'db>> {
+        // Extract `T` from a type annotation of the form `list[T]`.
+        fn annotated_elts_ty<'db>(
+            tcx: TypeContext<'db>,
+            collection_class: KnownClass,
+            db: &'db dyn Db,
+        ) -> Option<Type<'db>> {
+            let class_type = match tcx.annotation? {
+                Type::NominalInstance(instance) => instance,
+                Type::TypeAlias(alias) => alias.value_type(db).into_nominal_instance()?,
+                _ => return None,
+            }
+            .class(db);
+
+            if !class_type.is_known(db, collection_class) {
+                return None;
+            }
+            let specialization = class_type.into_generic_alias()?.specialization(db);
+            let [annotated_elts_ty] = specialization.types(db) else {
+                return None;
+            };
+
+            Some(*annotated_elts_ty)
+        }
+
+        // Extract the type variable `T` from `list[T]` in typeshed.
+        fn elts_ty(
+            collection_class: KnownClass,
+            db: &dyn Db,
+        ) -> Option<(ClassLiteral<'_>, Type<'_>)> {
+            let class_literal = collection_class.try_to_class_literal(db)?;
+            let generic_context = class_literal.generic_context(db)?;
+            let variables = generic_context.variables(db);
+            let elts_ty = variables.iter().exactly_one().ok()?;
+            Some((class_literal, Type::TypeVar(*elts_ty)))
+        }
+
+        let annotated_elts_ty = annotated_elts_ty(tcx, collection_class, self.db());
+        let (class_literal, elts_ty) = elts_ty(collection_class, self.db()).unwrap_or_else(|| {
+            let name = collection_class.name(self.db());
+            panic!("Typeshed should always have a `{name}` class in `builtins.pyi` with a single type variable")
+        });
+
+        let mut elements_are_assignable = true;
+        let mut inferred_elt_tys = Vec::with_capacity(elts.len());
+
+        // Infer the type of each element in the collection literal.
         for elt in elts {
-            self.infer_expression(elt, TypeContext::default());
+            let inferred_elt_ty = self.infer_expression(elt, TypeContext::default());
```

Comment:
Should we have a TODO here for pushing down the nested context? I saw some related TODOs in tests.

Reply:
I hadn't considered that supporting nested context was as simple as just passing the inner type down here! I implemented that and removed the TODOs.